dotnet/sdk-phase-1.md

```diff
@@ -0,0 +1,1101 @@
+
+* We use `System.Text.Json` by default and therefore support all customization/approaches it provides
+* Our JSON payload converters may offer easy approaches to customize options to create new converters
+
+### Dependency Injection
+
+https://learn.microsoft.com/en-us/dotnet/core/extensions/dependency-injection
+
+* We have `Temporalio.Client.ITemporalClient` and `Temporalio.Client.ITemporalConnection` to allow DI for these impls
+* Should we also do this for `Temporalio.Client.WorkflowHandle`? Is there value in injecting that?
+* Since activity instances are instantiated by the user upon registration, the user can use DI when creating those
+  classes to, say, inject a logger
+  * We'll need an example showing this
+* When we start the workflow implementation side, we may need to revisit how DI may help users without forcing it
+
+### Configuration
+
+https://learn.microsoft.com/en-us/dotnet/core/extensions/configuration
+
+* We support the https://learn.microsoft.com/en-us/dotnet/core/extensions/options pattern with all of our `Options`
+  classes
+  * Some of the options things like activities method references can't by loaded from a file, so we should take a look
+    at helpers for converting activity strings to activity classes reflectively
+  * We need a sample showing configuring workers via config file akin to Spring Boot
+* See the host section for how it all may come together inside a worker service
+
+### Logging
+
+https://learn.microsoft.com/en-us/dotnet/core/extensions/logging
+
+* Unfortunately there is no logging in the standard library for .NET, therefore in order to support this we'd have to
+  add a dependency on https://www.nuget.org/packages/Microsoft.Extensions.Logging
+* NuGet does not support optional dependencies
+* Should we add the dependency?
+  * For:
+    * It's pretty common, e.g. Azure durable task framework has it (well, `Microsoft.Extensions.Logging.Abstractions`)
+      but a really old version from 4 years ago
+  * Against:
+    * We don't like extra dependencies, even these tiny ones
+    * We will have a rudimentary `Temporalio.ITemporalLogger` with a simple console implementation by default
+      * Could have a separate `Temporalio.Extensions` project that had a `ITemporalLogger` impl for `ILogger`, but this
+        is so simple it can just be a sample
+  * Leaning towards "Against"
+
+### HostBuilder and Worker Services
+
+https://learn.microsoft.com/en-us/dotnet/core/extensions/generic-host and
+https://learn.microsoft.com/en-us/dotnet/core/extensions/workers
+
+* Same discussion around dependencies as logging, this would need `Microsoft.Extensions.Hosting`
+* Our `Temporalio.Worker.TemporalWorker` is not a subclass of `BackgroundService` to avoid the dependency
+  * `TemporalWorker.ExecuteAsync` matches `BackgroundService.ExecuteAsync` very intentionally, so this is incredibly
+    easy for a user to create and we will of course have a sample
```

Comment:
I think this one is more understandable than the logging one. I would wonder if it would be worth including in some sort of hosting extensions NuGet package though?

I've seen even reasonably experienced dotnet developers really struggle to understand how to:
- Properly handle graceful shutdowns
- Actively make decisions about how to handle exceptions in the service itself
- Sub-optimal blocking with `while(true)` type loops

If we we're to include a sane BackgroundService implementation somewhere I think it would go a long way to supporting the `Be as simple as possible from a caller POV` and `Be reasonably flexible with reasonable defaults` goals

Reply:
> I would wonder if it would be worth including in some sort of hosting extensions NuGet package though?

I considered that, but it'd be like a 3 line project. From a different section:

> Could have a separate `Temporalio.Extensions` project ...

Reply:
Cool, if the temporal worker implements some sort of StartAsync type method which has all the right handling of the cancellation token in it I think that's really good and there would be no need for a hosting package. i.e. something like:

```csharp
public Task StartAsync(CancellationToken stoppingToken)
{
    _temporalWorker.StartAsync(stoppingToken);
}
```

Reply:
This is the helper:

```csharp
public class WorkerBackgroundService : Microsoft.Extensions.Hosting.BackgroundService
{
  private readonly Worker worker;
  public WorkerBackgroundService(Worker worker) => this.worker = worker;
  protected override Task ExecuteAsync(CancellationToken stoppingToken) => worker.ExecuteAsync(stoppingToken);
}
```